### ISSUE TITLE
A Possible Improvement for Unreachable Code Detection about String Comparison in C #

### ISSUE THREAD

**Issue by vsfeedback** [dotnet/roslyn#50788]
_This issue has been moved from [a ticket on Developer Community](https://developercommunity2.visualstudio.com/t/A-Possible-Improvement-for-Unreachable-C/1316259)._

---
[Compiler warning CS0162] (https://docs.microsoft.com/zh-tw/dotnet/csharp/misc/cs0162) provides unreachable code detection feature. I think that there is a possible improvement for unreachable code detection.

Let's take a look at the first situation of string comparison like below:

```C#
const string test = "";
if (test. Equals("123"))
{
    //Do something...
}
```

Considering the following things:

- All values are constant in the if condition

- The determination of the if statement can be finished in compile time

The effect of the above `if` block is similar to `if (false) {  }`. Maybe this unreachable code block can be detected.

Next, let's take a look at another situation of string comparison:

```C#
string test = "";
if (test. Equals("123") && test. Equals("456"))
{
    //Do something...
}
```

Although a part of value (string `test`) is not constant in the if condition, no matter what content the string `test` is, `test` can't equal to "123" and "456" simultaneously. As the result, the if condition is always false.

---
#### Feedback Bot on 1/21/2021, 00:03 AM: 

Thank you for taking the time to provide your suggestion.  We will do some preliminary checks to make sure we can proceed further.  We'll provide an update once the issue has been triaged by the product team.

**Comment by Youssef1313** [comment]
While it's obvious that the code inside `if` is unreachable, the boolean expression of `if` is not a compile-time constant. Besides that, it would be a breaking change to introduce the warning for these.

However, this *may* be handled in roslyn-analyzers by CA1508.

@mavasani @Evangelink Do you think CA1508 should handle this?

**Comment by mavasani** [comment]
@Youssef1313 Agreed. CA1508 already covers the case where comparison is done with an operator, but doesn't seem to flag the case where Equals method is invoked:

![image](https://user-images.githubusercontent.com/10605811/105889495-14f82400-5fc3-11eb-99c0-97c11868923c.png)
